### PR TITLE
crypto: Storage changes for keeping sender data with InboundGroupSessions

### PR DESCRIPTION
**Pull request by andybalaam**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/3542

This changes the shape of `InboundGroupSession` to allow us to include a `SenderData` struct. To see how it will be populated, look at https://github.com/matrix-org/matrix-rust-sdk/issues/3543

### CODE REVIEW

**Comment by Hywan** [comment]
Is it ready for a first review?

**Comment by andybalaam** [comment]
> Is it ready for a first review?

No, just prototyping for now, thanks!

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+use vodozemac::Ed25519PublicKey;
+
+use crate::types::DeviceKeys;
+
+/// Information on the device and user that sent the megolm session data to us
+///
+/// Sessions start off in `UnknownDevice` state, and progress into `DeviceInfo`
+/// state when we get the device info. Finally, if we can look up the sender
+/// using the device info, the session can be moved into `SenderKnown` state.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:26]: If we sorted the derives we usually  sort these by groups, first come the traits from the standard library, then the rest. Each group is alphabetically sorted.

Not really important, and `cargo fmt` can't sort these for you so I don't want to enforce a style either, but perhaps you'd be interested in following an existing convention that is used at least in some places.

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+pub enum SenderData {
+    /// We have not yet found the (signed) device info for the sending device
+    UnknownDevice {
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:29]: It's a bit hard to know if these fields are what we want without seeing how they will be used, hopefully you got it right and won't have to fiddle with migrations :sweat_smile:

**Review by poljar — CHANGES_REQUESTED** [review]

File: .typos.toml [hunk 3]
```diff
@@ -36,4 +36,5 @@ extend-exclude = [
     "crates/matrix-sdk-ui/tests/integration/room_list_service.rs",
     # Hand-crafted base64 session keys that are understood as typos.
     "crates/matrix-sdk-indexeddb/src/crypto_store/migrations/mod.rs",
+    "crates/matrix-sdk-crypto/src/olm/group_sessions/inbound.rs",
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply .typos.toml:39]: Oh, I was seeing it as a group: this one is the same reason as the one above. Happy to add a separate comment for clarity though.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 4]
```diff
@@ -220,6 +221,7 @@ impl StaticAccountData {
             signing_key,
             room_id,
             &outbound.session_key().await,
+            SenderData::unknown(),
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/account.rs:224]: Yeah, I am addressing this in https://github.com/matrix-org/matrix-rust-sdk/pull/3590

File: crates/matrix-sdk-crypto/src/types/events/room_key.rs [hunk 5]
```diff
@@ -113,15 +113,22 @@ pub struct MegolmV1AesSha2Content {
     ///
     /// [`InboundGroupSession`]: vodozemac::megolm::InboundGroupSession
     pub session_key: SessionKey,
+    /// The device keys if supplied as per MSC4147
+    pub device_keys: Option<DeviceKeys>,
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/types/events/room_key.rs:117]: Ah right, yeah I need to adapt this now that https://github.com/matrix-org/matrix-rust-sdk/pull/3517 has merged - hopefully it does this stuff for me.

File: crates/matrix-sdk-crypto/src/types/events/room/encrypted.rs [hunk 6]
```diff
@@ -431,6 +431,7 @@ pub(crate) mod tests {
                 "device_id": "DEWRCMENGS",
                 "session_id": "ZFD6+OmV7fVCsJ7Gap8UnORH8EnmiAkes8FAvQuCw/I",
                 "sender_key": "WJ6Ce7U67a6jqkHYHd8o0+5H4bqdi9hInZdk0+swuXs",
+                "device_keys": null,
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/types/events/room/encrypted.rs:434]: This is used for round-tripping I think, so it was needed.

File: crates/matrix-sdk-crypto/src/olm/group_sessions/inbound.rs [hunk 7]
```diff
@@ -611,7 +648,134 @@ mod tests {
...
+
+        // Then it was parsed correctly
+        assert_eq!(unpickled.session_id(), "XbmrPa1kMwmdtNYng1B2gsfoo8UtF+NklzsTZiaVKyY");
+
+        // And we populated the InboundGroupSession's sender_data with a default value,
+        // with legacy_session set to true.
+        let SenderData::UnknownDevice { retry_details, legacy_session } = unpickled.sender_data
+        else {
+            panic!("Expected sender_data to be UnknownDevice!");
+        };
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/inbound.rs]: Fixed in d6a19f36c3

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs]: Definitely. Fixed in a39f7a5e85

File: crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs (hunk 2 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs:29]: Yes, I know and maybe I should have done this all together with https://github.com/matrix-org/matrix-rust-sdk/pull/3590 but it just seemed too big.

I did wait until I had https://github.com/matrix-org/matrix-rust-sdk/pull/3590 working before I set this ready to review, so we have at least some chance that I got it right.

File: .typos.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (poljar) [reply .typos.toml:39]: In that case, please add the individual base64 false positives to the `default.extend-words` section. The `inbound.rs` file contains enough docs for which we shouldn't disable typo checking.

**Review by poljar — APPROVED** [review]

File: .typos.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (andybalaam) [reply .typos.toml:39]: Done in https://github.com/matrix-org/matrix-rust-sdk/pull/3629

File: crates/matrix-sdk-crypto/src/types/events/room_key.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/types/events/room_key.rs:117]: The fix is https://github.com/matrix-org/matrix-rust-sdk/pull/3633

**Pull request merged**